include/bredis/AsyncConnection.hpp

```diff
@@ -0,0 +1,69 @@
+//
+//
+// Copyright (c) 2017 Ivan Baidakou (basiliscos) (the dot dmol at gmail dot com)
+//
+// Distributed under the MIT Software License
+//
+
+#pragma once
+
+#include <atomic>
+#include <initializer_list>
+#include <mutex>
+#include <queue>
+#include <sstream>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include <boost/asio.hpp>
+#include <boost/utility/string_ref.hpp>
+
+#include "Command.hpp"
+#include "Error.hpp"
+#include "Protocol.hpp"
+#include "Result.hpp"
+
+namespace bredis {
+
+template <typename AsyncStream> class AsyncConnection {
+    using protocol_type_t = typename AsyncStream::protocol_type;
+
+    static_assert(std::is_same<protocol_type_t, boost::asio::ip::tcp>::value ||
+                      std::is_same<protocol_type_t,
+                                   boost::asio::local::stream_protocol>::value,
+                  "AsyncConnection can be specialized either for ip::tcp or "
+                  "local::stream_protocol");
+
+  public:
+    using string_t = boost::string_ref;
+    using args_container_t = std::vector<string_t>;
+
+  private:
+    AsyncStream socket_;
+
+  public:
+    template <typename... Args>
+    AsyncConnection(Args &&... args) : socket_(std::forward<Args>(args)...) {}
+
+    inline AsyncStream &next_layer() { return socket_; }
+    inline AsyncStream &&move_layer() { return std::move(socket_); }
```

Comment:
This function is going to cause problems. First of all, there's no way to guarantee that `AsyncStream` is movable. You will have a function called `move_layer` which in some cases will perform a copy. Or worse, do nothing. For example what if `AsyncStream` is a reference type? The function would be meaningless in that case. I would just leave out this function. If callers don't want the connection object to have ownership of the socket they can just declare it with a reference type instead.